main/lsp/LSPIndexer.cc

```diff
@@ -10,86 +10,24 @@
-            ++i;
-            auto fref = gs.findFileByPath(f->path());
-            if (!fref.exists()) {
-                logger.debug("Taking slow path because {} is a new file", f->path());
-                prodCategoryCounterInc("lsp.slow_path_reason", "new_file");
-                return false;
-            } else {
-                auto &oldHash = findHash(fref.id(), globalStateHashes, overriddingStateHashes);
-                ENFORCE(oldHash.definitions.hierarchyHash != core::GlobalStateHash::HASH_STATE_NOT_COMPUTED);
-                if (changedHashes[i].definitions.hierarchyHash == core::GlobalStateHash::HASH_STATE_INVALID) {
-                    logger.debug("Taking slow path because {} has a syntax error", f->path());
-                    prodCategoryCounterInc("lsp.slow_path_reason", "syntax_error");
-                    return false;
-                } else if (changedHashes[i].definitions.hierarchyHash != core::GlobalStateHash::HASH_STATE_INVALID &&
-                           changedHashes[i].definitions.hierarchyHash != oldHash.definitions.hierarchyHash) {
-                    logger.debug("Taking slow path because {} has changed definitions", f->path());
-                    prodCategoryCounterInc("lsp.slow_path_reason", "changed_definition");
-                    return false;
-                }
-            }
-        }
+const core::File &getOldFile(core::FileRef fref, const core::GlobalState &gs,
+                             const UnorderedMap<int, shared_ptr<core::File>> &evictedFiles) {
+    const auto &it = evictedFiles.find(fref.id());
+    if (it != evictedFiles.end()) {
+        return *it->second;
     }
-    logger.debug("Taking fast path");
-    return true;
-}
-
-bool updateCanTakeFastPath(
-    const core::GlobalState &gs, const LSPConfiguration &config, const vector<core::FileHash> &globalStateHashes,
-    const LSPFileUpdates &updates,
-    const UnorderedMap<int, core::FileHash> &overriddingStateHashes = UnorderedMap<int, core::FileHash>()) {
-    return canTakeFastPathInternal(gs, config, globalStateHashes, updates.updatedFileHashes, updates.updatedFiles,
-                                   updates.hasNewFiles, overriddingStateHashes);
+    ENFORCE(fref.exists());
+    return fref.data(gs);
 }
 
-UnorderedMap<int, core::FileHash> mergeEvictions(const UnorderedMap<int, core::FileHash> &olderEvictions,
-                                                 const UnorderedMap<int, core::FileHash> &newerEvictions) {
-    // For evictions, which are needed for emulating an older `globalStateHashes`, we keep the oldest.
-    UnorderedMap<int, core::FileHash> combinedEvictions = olderEvictions;
-    for (auto &e : newerEvictions) {
-        if (!combinedEvictions.contains(e.first)) {
-            combinedEvictions[e.first] = e.second;
-        }
+// Merges *oldEvictedFiles* into *newlyEvictedFiles*. Mutates newlyEvictedFiles.
+void mergeEvictedFiles(const UnorderedMap<int, shared_ptr<core::File>> &oldEvictedFiles,
```

Comment:
Replacement for `mergeEvictions`, except it mutates an argument to reduce copying.